### ISSUE TITLE
Gmail/Inbox support (with sample code)

### ISSUE THREAD

**Issue by pjf** [ralphbean/bugwarrior#442]
## Desired feature

I use Google's Inbox to help keep on top of tasks that come in via email, and I would *love* for this to be integrated into TaskWarrior. In particular, there are two common workflows that I know would be very useful:

* Users who 'pin' and 'snooze' their messages in inbox.
* Users who 'star' messages in their gmail

## The Gmail API

The Gmail API allows us to work with both of these workflows, and I'll briefly cover my discoveries thus far.

* Google has a [python quickstart](https://developers.google.com/gmail/api/quickstart/python) example that shows the user's folders. It requires creating an API token and secret, which is relatively straightforward.
* To search for inbox pinned messages, one can modify the `results` line of the quickstart to `results = service.users().threads().list(userId='me',q="in:pinned").execute()` 
* To search for inbox snoozed messages, one can use `results = service.users().threads().list(userId='me',q="label:snoozed").execute()`
* To search for starred messages, one should be able to use ``results = service.users().threads().list(userId='me',q="is:starred").execute()` (haven't personally tested)

All of these return a [list of threads](https://developers.google.com/gmail/api/v1/reference/users/threads), although for starred items a [list of messages](https://developers.google.com/gmail/api/v1/reference/users/messages) probably makes more sense. I haven't done the assembly required to extract the subject lines and relevant message bodies, but it looks like a Simple Matter Of Code.

## Behaviour in BugWarrior/TaskWarrior

For folks using pinned/snoozed threads, I'd have the following behaviour:

* Pinned threads are turned into active tasks in TaskWarrior.
* Snoozed threads are turned into waiting tasks in TaskWarrior.

I don't know if the API gives us any way to determine how long a task is snoozed for, so we may just mark them as waiting forever, and then remove that status when we see them revert to pinned.

For starred messages (which I don't personally use) I'm guessing we want all starred messages in a given list of folders to become tasks. By default that's probably just the inbox, but we should allow users with more advanced habits to specify a folder list.

## Resources I can provide

If this is considered within the scope of bugwarrior, then I'm delighted to help where I can. Normally I'd do that by sending in code, but while I've got decades of experience writing software, my python skills are currently poor, so we'd have to deal with my extended learning time. So realistically I can provide testing and minor bugfixes, and to be boundlessly enthusiastic to my python-wielding friends about how great such a feature would be and how we've already got the relevant APIs calls to make it happen. :)

Many, many thanks again for bugwarrior,

~ Paul

**Comment by ryneeverett** [comment]
Sounds reasonable to me.

This isn't a blocker on working on this, but I'd like to suggest that it would be helpful to get more feedback on how people use gmail for task management. Your two suggested workflows sound right to me, but it's more reliable to ask how people use software than to try to guess. **If you're reading this and you use gmail for task management, tell us about it!**

> If this is considered within the scope of bugwarrior, then I'm delighted to help where I can.

This is grey territory. We haven't ventured into it yet but I think we're interested. The main distinction in my mind is not that gmail isn't a bug tracker, it's that the task workflow is much more subjective. However, your proposal suggests that maybe there are only a couple reasonable ways to use gmail for task management. If it turns out that everybody uses gmail differently for this purpose, gmail is not a good fit for bugwarrior because we can only support so many use cases.

> while I've got decades of experience writing software, my python skills are currently poor, so we'd have to deal with my extended learning time.

Perhaps you'll recruit an interested pythonista, but if not this is a good project to learn on.

**Comment by pjf** [comment]
@ryneeverett : Agreeing with you on everything! And I can absolutely encourage reports on how people use gmail for task management. :)

> Perhaps you'll recruit an interested pythonista, but if not this is a good project to learn on.

It turns out I have a client who's supporting my learning to python, so there's a chance I'll be doing both! :)

**Comment by pjf** [comment]
From a configuration standpoint, I think we definitely want our users to be able to specify their own query strings, as that lets them do things like `in:inbox label:read` so all *read* email can move into taskwarrior, but unread email gets ignored. Gmail queries can get [quite expressive](https://support.google.com/mail/answer/7190), so this is likely to give us a lot of coverage of workflows where users are combining labels, pinnning, snoozing, stars of different colours, senders, recipients, and all other manner of things.

I'm still learning what BugWarrior can and cannot do, and so my config format may be wrong here, but something like this would be amazing:

```
[gmail-pinned]
# Pinned msgs aren't waiting, they're actionable.
query = in.pinned
wait_template =

[gmail-snoozed]
# Snoozed msgs are 'waiting'.
query = label:snoozed
wait_template = +7d

[gmail-complex]
# Some important things need to be handled immediately
query = in:alerts from:emergency@example.com subject:fire
default_priority = H
add_tags = Emergency
```

**Comment by ryneeverett** [comment]
Cool, so we may only need to integrate with the query API.

**Comment by emmajane** [comment]
Hello! I am reading this and I use Gmail for task management! I specifically use their keyboard shortcut to send an email to their "tasks".

My workflow currently looks like this:

- Triage Gmail inbox using `[` and `]` shortcuts (send to archive; move to next message in the inbox).
- If there is an email that requires follow-up, I first send the message to Gmail tasks using `shift+t`. 

This action adds the task to a pop-up window which appears in the bottom right of my screen (see attached).

![inbox_-_emma_humanitarianresponse_info_-_humanitarianresponse_info_mail](https://cloud.githubusercontent.com/assets/1299320/23262675/6e49967e-f9d3-11e6-8dce-32a34d665188.png)

This also appears as a stand-alone screen at: https://mail.google.com/tasks/canvas?pli=1

There are Gmail help pages at: https://support.google.com/mail/answer/106237?hl=en I believe this is the relevant API documentation: https://developers.google.com/google-apps/tasks/

I'm happy to provide more details if they would be useful.

**Comment by ryneeverett** [comment]
Thanks @emmajane. Somehow we overlooked google's actual task management system.

@pjf I'm going to add a section to your initial post to track the current status of this issue since this thread could become long.

**Comment by cthulahoops** [comment]
This is my attempt at this:

https://github.com/cthulahoops/bugwarrior/tree/issue_442_add_gmail_service

Seems to be working fine. I just need to clean up and figure out the testing infrastructure and it should be ready to open a pull request.

(Adding google tasks should be no harder, but I haven't been using them.)

**Comment by ryneeverett** [comment]
@cthulahoops Looking quite good at a cursory glance. Please open a pull request. You can call it a "[WIP]" if there are still TODO items but as far as I can tell it's ready for review.

**Comment by ryneeverett** [comment]
Initial issue by @pjf resolved by #477. Google tasks integration as suggested by @emmajane is spun off to #487.
